module/dkg/recovery.go

```diff
@@ -0,0 +1,155 @@
+		state:         state,
+		localDKGState: localDKGState,
+	}
+
+	err := recovery.tryRecoverMyBeaconPrivateKey(state.Final())
+	if err != nil {
+		return nil, fmt.Errorf("could not recover my beacon private key when initializing: %w", err)
+	}
+
+	return recovery, nil
+}
+
+// EpochFallbackModeExited implements handler from protocol.Consumer to perform recovery of the beacon private key when
+// this node has exited the epoch fallback mode.
+func (b *BeaconKeyRecovery) EpochFallbackModeExited(epochCounter uint64, refBlock *flow.Header) {
+	b.log.Info().Msgf("epoch fallback mode exited for epoch %d", epochCounter)
+	err := b.tryRecoverMyBeaconPrivateKey(b.state.AtHeight(refBlock.Height)) // refBlock must be finalized
+	if err != nil {
+		irrecoverable.Throw(context.TODO(), fmt.Errorf("failed to get final epoch protocol state: %w", err))
+	}
+}
+
+// tryRecoverMyBeaconPrivateKey performs the recovery of the random beacon private key for the next epoch by trying to use
+// a safe 'my beacon key' from the current epoch (it is expected that this method will be called before entering the recovered epoch).
+// If a safe 'my beacon key' is found, it will be stored in the storage.EpochRecoveryMyBeaconKey for the next epoch
+// concluding the 'my beacon key' recovery.
+// If there is a safe 'my beacon key' for the next epoch, or we are not in committed phase (DKG for next epoch is not available)
+// then calling this method is no-op.
+// No errors are expected during normal operations.
+func (b *BeaconKeyRecovery) tryRecoverMyBeaconPrivateKey(final protocol.Snapshot) error {
+	head, err := final.Head()
+	if err != nil {
+		return fmt.Errorf("could not get head of snapshot: %w", err)
+	}
+	epochProtocolState, err := final.EpochProtocolState()
+	if err != nil {
+		return fmt.Errorf("could not get epoch protocol state: %w", err)
+	}
+	currentEpochCounter := epochProtocolState.Epoch()
+
+	log := b.log.With().
+		Uint64("height", head.Height).
+		Uint64("view", head.View).
+		Uint64("epochCounter", currentEpochCounter).
+		Logger()
+	if epochProtocolState.EpochPhase() != flow.EpochPhaseCommitted {
+		log.Info().Msgf("epoch is in phase %s", epochProtocolState.EpochPhase())
+		return nil
+	}
+
+	nextEpochCounter, err := final.Epochs().Next().Counter()
+	if err != nil {
+		return fmt.Errorf("could not get next epoch counter: %w", err)
+	}
+	_, safe, err := b.localDKGState.RetrieveMyBeaconPrivateKey(nextEpochCounter)
+	if err != nil && !errors.Is(err, storage.ErrNotFound) {
+		return fmt.Errorf("could not retrieve my beacon private key for the next epoch: %w", err)
+	}
+	if safe {
+		log.Info().Msg("my beacon private key for the next epoch is safe, nothing to do")
+		return nil
+	}
+
+	myBeaconPrivateKey, safe, err := b.localDKGState.RetrieveMyBeaconPrivateKey(currentEpochCounter)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			log.Warn().Str(logging.KeySuspicious, "true").Msgf("no my beacon key for the current epoch has been found")
```

Comment:
I agree with not using `KeySuspicious` here, and with extending its documentation -- I suspect it is currently being used overly broadly because of the vague documentation ([example](https://github.com/onflow/flow-go/blob/01fcf31d1eb7952c0e0276d8c386c5a5d57dd715/network/p2p/builder/gossipsub/gossipSubBuilder.go#L69)). Adding a logging key for configuration problems makes sense to me. 

(By the way, there is also `KeyProtocolViolation` for _certain_ Byzantine behaviour.)

Reply:
[`b334c81` (#6424)](https://github.com/onflow/flow-go/pull/6424/commits/b334c811f5c88941e7a1fa3a7e067c4a5e291350)